Add deserializeMutated test for datastructures

jtraglia opened the pull request:
## PR Description

The current property tests deserialize input that should be deserializable. This PR adds tests that deserialize data structures that have been slightly mutated. I believe this is smarter than deserializing random bytes. This PR is focussed on SSZ deserialization; we could do the same thing for JSON deserialization, but it would be a bit more difficult.

Overview of changes:

* Add new `Mutator` class with a static `mutate` method.
* Add new `deserializeMutated` method to each data structure property test file.
* Add new `assertDeserializeMutatedThrowsExpected` method to `PropertyTestHelper`.
  * Expected exceptions are `SszDeserializeException` and `IllegalArgumentException`.

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

jtraglia committed:
Add deserializeMutated test for datastructures

tbenr commented on ethereum/spec/src/property-test/java/tech/pegasys/teku/spec/datastructures/util/Mutator.java:
```diff
@@ -0,0 +1,131 @@
+      if (randomNumber <= MUTATION_PROBABILITY) {
+
+        Mutation op = null;
+        for (final Map.Entry<Mutation, Range> entry : RANGE_MAP.entrySet()) {
```
You might use a `TreeMap<Double,Mutation>.floorEntry` and you don't need a `Range` but just the upper bound of the probability

jtraglia replied:
Good idea. I'll look into that!

jtraglia replied:
Really great suggestion. Thanks. I think this looks a lot better. I ended up using `TreeMap.ceilingEntry` instead of `floorEntry` because calling `floorEntry` with a number greater than the largest key would return the last entry. I think `ceilingEntry` what you meant to suggest anyway, right?

jtraglia committed:
Change mutation probability check

Technically, I think this should be "less than" and not "less than or equal to"
because contains() checks the exclusive range.

tbenr reviewed (commented):
Looks Cool.

jtraglia committed:
Update Mutator to use TreeMap.ceilingEntry

jtraglia committed:
Merge branch 'master' into deserialize-mutated

ajsutton committed:
Use ByteArrayOutputStream to make mutations faster.

ajsutton committed:
Merge branch 'master' of github.com:ConsenSys/teku into deserialize-mutated

ajsutton reviewed (approved):
LGTM. Thanks.

I've pushed one little tweak to how the mutation is done so that it uses `ByteArrayOutputStream` rather than having to create so many individual `Bytes` instances. It probably helps performance but also just reads a bit more naturally.
